### DISCUSSION TITLE
What's going on with the iterate-function (Chapter 4, page103)

### DISCUSSION THREAD

**Discussion by PhilippAubert** [PacktPublishing/Node.js-Design-Patterns-Third-Edition#50] [Q&A]
Hello everyone - I have a very simple question about the "pattern" of that recursive iteration. 
I understand it's quite powerful - but when trying to work this code, it's throwing an error. 
```

function iterate(index) {
  if (index === tasks.length) {
    return finish;
  }
  const task = tasks[index];
  task(() => iterate(index + 1));
}

function finish() {
  console.log("finished");
}

iterate(0);
```

I get the error : `TypeError: task is not a function` 

Maybe I'm too noob for this ;D

**Comment by PhilippAubert** [comment]
ok, so the main thing seems to be this: 
```

  const task = tasks[index];
  task(() => iterate(index + 1));
```

how can this const become a function one line later?

**Reply by mariocasciaro** [reply]
It's because `tasks` must be an array of functions (tasks, in fact). For example:

```
const tasks = [
  () => {...},
  () => {...}
]
```

But you can easily adjust the pattern to accept a list of objects and pass them to a function

**Reply by PhilippAubert** [reply]
Makes sense. I tried to write a very simple array from 1-10 which didn't do the trick.

**Reply by PhilippAubert** [reply]
Thank you!!!
